### ISSUE TITLE
Mobile Navbar not closing on Navigation

### ISSUE THREAD

**Issue by pfried** [tailwindlabs/tailwindui-issues#572]
**What component (if applicable)**
- URL for category: 
- https://tailwindui.com/components/application-ui/navigation/navbar
- Component name: 
- Simple

**Describe the bug**
I asked on discord, and I have not been the only one how how to solve this. Since there seems to be no easy answer I wanted to create an issue here:

This applies to the mobile navigation with the hamburger / close icon on small screens.

There is an expectation that a Navbar would close when you click a link within the component. E.g. you have a SPA react application and you navigate internally without page reloads (so no real `<a href="...">` links). Especially on mobile devices you will click a link, but cannot see the content due to the overlapping menu. You do not get feedback on the navigation happening and you will have to click the close icon on top right first.

**To Reproduce**
1. Go to the Navbar component and narrow it down so that the mobile menu appears
2. Open the mobile navbar by clicking the hamburger icon
3. Select on of the options in the navbar
4. The navbar does not close

**Expected behavior**
I understand that some items do need the menu to stay open but it would be great if the navbar (the underlying `Disclosure` component exposes a method to close the menu.

Or the Navbar component is rebuild on the `Menu` component so that it automatically closes when you select an item (just like the user navigation in the Navbar component)

**Screenshots**
![image](https://user-images.githubusercontent.com/1744875/115183485-23df0600-a0dc-11eb-98eb-ad97aebc6229.png)

**Comment by capofalcone** [comment]
Also I asked on discord the same, but I'd like to add that I think Navbar should use the Menu Headless UI component and not the Disclosure component. As the Disclosure.Panel is used as a menu would make much more sense use the Menu component.
For the Disclosure Headless UI would be useful if I we could pass an open prop to control the open state from outside the component, as it works for the Dialog component.

**Comment by jasonacat** [comment]
I am running into this same issue - mobile menu does not seem to actually function as a menu unless it causes a navigation away from the page. Why was a Disclosure used for mobile menu, even with 'Simple dark with menu button on left' where it doesn't have the fancy content? I am thinking of just rewriting to use a single Menu restyled for mobile, but I don't have the context on the component choices to know if there's some kind of accessibility reason for using a Disclosure on mobile.

**Comment by davidluhr** [comment]
Thanks for all this feedback. There are a few considerations to dig into here.

1. The `Disclosure` component is used to show/hide content, and the revealed content is render inline, shifting other content beneath it.
2. The `Menu` component renders the menu as an overlay, obscuring other content underneath.
3. These different visual treatments have accessibility considerations for keyboard navigation and assistive technology support. Any time content is obscured underneath an element, we need to properly manage focus to ensure that users can't accidentally move focus to content that is hidden, which can be very disorienting.
4. Since the visual design of some of these mobile navigation examples has the navigation rendered inline, `Disclosure` is the best fit for this expanding/collapsing functionality.
5. Disclosures can contain arbitrary content. Sometimes they just reveal text, and sometimes they contain interactive items. In this case, the Disclosure mostly contains links. By design, Disclosures don't close unless the user explicitly closes them (by clicking the toggle button), since they don't obscure other content.
6. Now, typically, clicking a link would result in a page refresh, but SPAs intervene in this behavior, as you're familiar with. The issue is that most SPA routers are inaccessible, as they don't manage focus properly on route change, and nothing is announced to assistive technology when the route changes.
7. As a result, we're left with an interesting scenario. Our implementation is correct, but we're faced with the limitations of SPA routing. We're considering an API update in Headless UI that allows users to opt into closing the Disclosure automatically in a situation like this as a workaround, but it is tricky.

Long story short, the existing implementation is actually correct, but obviously it reveals limitations in how some frameworks behave. So, we are thinking of ways to overcome these limitations, even though it's not necessarily a valid approach, to somewhat fake what would actually happen when clicking a real link and loading a new page.

Thanks for putting up with this wall of text. I don't say any of this to discourage this feedback, and in fact, this is something we're actively working to improve. I hope this is helpful, however, in understanding the gap between what's correct and how some frameworks choose to handle things. I look forward to any other thoughts you have, and I will keep this issue updated with any developments.

**Comment by pfried** [comment]
@davidluhr Thank you for the detailled feedback. The issue is understood, but since this is the tailwind ui issues repo and not the headless ui one I would propse changing the react examples to use a `state` which controlls the open/close state of the `Disclosure` That is at least how I solved this for now. Declaring the `Disclosure.Panel` as static and then doing a `onClick={() => setMobileOpen(false)}` on the items as required.

I know the examples get autocreated, so this might be an issue and also does not feel very good to do. But for a SPA the `Navbar` not closing is a bigger issue for me.

**Comment by reinink** [comment]
Just an update on the issue. Since originally posting this issue, Headless UI has been updated to allow using a `<Disclosure.Button>` within a `<Disclosure.Panel>` for the purpose of closing the disclosure when you click a link. You can read more about that [here](https://headlessui.dev/react/disclosure#closing-disclosures-manually).

Here's how:

```diff
 <Disclosure.Panel className="sm:hidden">
   <div className="px-2 pt-2 pb-3 space-y-1">
     {navigation.map((item) => (
-      <a
+      <Disclosure.Button
+        as="a"
         key={item.name}
         href={item.href}
         className={classNames(
           item.current ? 'bg-gray-900 text-white' : 'text-gray-300 hover:bg-gray-700 hover:text-white',
           'block px-3 py-2 rounded-md text-base font-medium'
         )}
         aria-current={item.current ? 'page' : undefined}
       >
         {item.name}
+      </Disclosure.Button>
-      </a>
     ))}
   </div>
 </Disclosure.Panel>
```

I am going to leave this issue open for now as a reminder to update all the Tailwind UI components to reflect this change.

**Comment by reinink** [comment]
Hey everyone, just wanted to let you know that we've updated all the Tailwind UI components to include a `<Disclosure.Button>` by default, so this shouldn't be an issue anymore! 🙌

Thanks for your patience, and thanks again to @pfried for reporting this 🙏
